File: sdk/csharp/TSKParser.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;

namespace TuskLang
{
    /// <summary>
    /// Parser for TSK (TuskLang Configuration) format
    /// Handles TOML-like syntax with sections, key-value pairs, and multiline strings
    /// </summary>
    public class TSKParser
    {
        /// <summary>
        /// Parse TSK content into dictionary
        /// </summary>
        public static Dictionary<string, object> Parse(string content)
        {
            var (data, _) = ParseWithComments(content);
            return data;
        }

        /// <summary>
        /// Parse TSK content with comments preserved
        /// </summary>
        public static (Dictionary<string, object> data, Dictionary<int, string> comments) ParseWithComments(string content)
        {
            var lines = content.Split('\n');
            var result = new Dictionary<string, object>();
            var comments = new Dictionary<int, string>();
            string currentSection = null;
            bool inMultilineString = false;
            string multilineKey = null;
            var multilineContent = new List<string>();

            for (int i = 0; i < lines.Length; i++)
            {
                var line = lines[i];
                var trimmedLine = line.Trim();

                // Handle multiline strings
                if (inMultilineString)
                {
                    if (trimmedLine == "\"\"\"")
                    {
                        if (currentSection != null && multilineKey != null)
                        {
                            if (!result.ContainsKey(currentSection))
                                result[currentSection] = new Dictionary<string, object>();
                            
                            if (result[currentSection] is Dictionary<string, object> sectionData)
                                sectionData[multilineKey] = string.Join("\n", multilineContent);
                        }
                        inMultilineString = false;
                        multilineKey = null;
                        multilineContent.Clear();
                        continue;
                    }
                    multilineContent.Add(line);
                    continue;
                }

                // Capture comments
                if (trimmedLine.StartsWith("#"))
                {
                    comments[i] = trimmedLine;
                    continue;
                }

                // Skip empty lines
                if (string.IsNullOrEmpty(trimmedLine))
                    continue;

                // Section header
                var sectionMatch = Regex.Match(trimmedLine, @"^\[(.+)\]$");
                if (sectionMatch.Success)
                {
                    currentSection = sectionMatch.Groups[1].Value;
                    result[currentSection] = new Dictionary<string, object>();
                    continue;
                }

                // Key-value pair
                if (currentSection != null && trimmedLine.Contains("="))
                {
                    var separatorIndex = trimmedLine.IndexOf('=');
                    var key = trimmedLine.Substring(0, separatorIndex).Trim();
                    var valueStr = trimmedLine.Substring(separatorIndex + 1).Trim();

                    // Check for multiline string start
                    if (valueStr == "\"\"\"")
                    {
                        inMultilineString = true;
                        multilineKey = key;
                        continue;
                    }

                    var value = ParseValue(valueStr);
                    if (result[currentSection] is Dictionary<string, object> sectionData)
                        sectionData[key] = value;
                }
            }

            return (result, comments);
        }

        /// <summary>
        /// Parse a TSK value string into appropriate C# type
        /// </summary>
        private static object ParseValue(string valueStr)
        {
            // Null
            if (valueStr == "null") return null;

            // Boolean
            if (valueStr == "true") return true;
            if (valueStr == "false") return false;

            // Number
            if (int.TryParse(valueStr, out int intValue)) return intValue;
            if (double.TryParse(valueStr, out double doubleValue)) return doubleValue;

            // String
            if (valueStr.StartsWith("\"") && valueStr.EndsWith("\""))
            {
                return valueStr.Substring(1, valueStr.Length - 2)
                    .Replace("\\\"", "\"")
                    .Replace("\\\\", "\\");
            }

            // Array
            if (valueStr.StartsWith("[") && valueStr.EndsWith("]"))
            {
                var arrayContent = valueStr.Substring(1, valueStr.Length - 2).Trim();
                if (string.IsNullOrEmpty(arrayContent)) return new List<object>();

                var items = SplitArrayItems(arrayContent);
                return items.Select(item => ParseValue(item.Trim())).ToList();
            }

            // Object/Dictionary
            if (valueStr.StartsWith("{") && valueStr.EndsWith("}"))
            {
                var objContent = valueStr.Substring(1, valueStr.Length - 2).Trim();
                if (string.IsNullOrEmpty(objContent)) return new Dictionary<string, object>();

                var pairs = SplitObjectPairs(objContent);
                var obj = new Dictionary<string, object>();

                foreach (var pair in pairs)
                {
                    var eqIndex = pair.IndexOf('=');
                    if (eqIndex > -1)
                    {
                        var key = pair.Substring(0, eqIndex).Trim();
                        var value = pair.Substring(eqIndex + 1).Trim();
                        // Remove quotes from key if present
                        var cleanKey = key.StartsWith("\"") && key.EndsWith("\"") 
                            ? key.Substring(1, key.Length - 2) 
                            : key;
                        obj[cleanKey] = ParseValue(value);
                    }
                }

                return obj;
            }

            // Return as string if no other type matches
            return valueStr;
        }

        /// <summary>
        /// Split array items considering nested structures
        /// </summary>
        private static List<string> SplitArrayItems(string content)
        {
            var items = new List<string>();
            var current = "";
            var depth = 0;
            var inString = false;

            for (int i = 0; i < content.Length; i++)
            {
                var ch = content[i];

                if (ch == '"' && (i == 0 || content[i - 1] != '\\'))
                {
                    inString = !inString;
                }

                if (!inString)
                {
                    if (ch == '[' || ch == '{') depth++;
                    if (ch == ']' || ch == '}') depth--;

                    if (ch == ',' && depth == 0)
                    {
                        items.Add(current.Trim());
                        current = "";
                        continue;
                    }
                }

                current += ch;
            }

            if (!string.IsNullOrEmpty(current.Trim()))
            {
                items.Add(current.Trim());
            }

            return items;
        }

        /// <summary>
        /// Split object pairs considering nested structures
        /// </summary>
        private static List<string> SplitObjectPairs(string content)
        {
            var pairs = new List<string>();
            var current = "";
            var depth = 0;
            var inString = false;

            for (int i = 0; i < content.Length; i++)
            {
                var ch = content[i];

                if (ch == '"' && (i == 0 || content[i - 1] != '\\'))
                {
                    inString = !inString;
                }

                if (!inString)
                {
                    if (ch == '[' || ch == '{') depth++;
                    if (ch == ']' || ch == '}') depth--;

                    if (ch == ',' && depth == 0)
                    {
                        pairs.Add(current.Trim());
                        current = "";
                        continue;
                    }
                }

                current += ch;
            }

            if (!string.IsNullOrEmpty(current.Trim()))
            {
                pairs.Add(current.Trim());
            }

            return pairs;
        }

        /// <summary>
        /// Convert dictionary back to TSK string format
        /// </summary>
        public static string Stringify(Dictionary<string, object> data)
        {
            var sb = new StringBuilder();

            foreach (var section in data)
            {
                sb.AppendLine($"[{section.Key}]");

                if (section.Value is Dictionary<string, object> sectionData)
                {
                    foreach (var kvp in sectionData)
                    {
                        sb.AppendLine($"{kvp.Key} = {FormatValue(kvp.Value)}");
                    }
                }
                else
                {
                    sb.AppendLine($"value = {FormatValue(section.Value)}");
                }

                sb.AppendLine();
            }

            return sb.ToString().TrimEnd();
        }

        /// <summary>
        /// Format a value for TSK string representation
        /// </summary>
        private static string FormatValue(object value)
        {
            if (value == null) return "null";
            if (value is bool b) return b.ToString().ToLower();
            if (value is string str) return $"\"{str.Replace("\"", "\\\"").Replace("\\", "\\\\")}\"";
            if (value is int || value is long || value is double || value is float) return value.ToString();

            if (value is Dictionary<string, object> dict)
            {
                var pairs = dict.Select(kvp => $"\"{kvp.Key}\" = {FormatValue(kvp.Value)}");
                return $"{{{string.Join(", ", pairs)}}}";
            }

            if (value is IEnumerable<object> enumerable)
            {
                var items = enumerable.Select(FormatValue);
                return $"[{string.Join(", ", items)}]";
            }

            if (value is IEnumerable<KeyValuePair<string, object>> kvpEnumerable)
            {
                var pairs = kvpEnumerable.Select(kvp => $"\"{kvp.Key}\" = {FormatValue(kvp.Value)}");
                return $"{{{string.Join(", ", pairs)}}}";
            }

            return $"\"{value}\"";
        }
    }
} 

File: sdk/ruby/tsk_parser.rb
# frozen_string_literal: true

# TSK Parser for Ruby
# Handles TOML-like TSK format parsing and generation

module TuskLang
  # Parser for TSK (TuskLang Configuration) format
  class TSKParser
    # Parse TSK content into hash
    def self.parse(content)
      data, _ = parse_with_comments(content)
      data
    end

    # Parse TSK content with comments preserved
    def self.parse_with_comments(content)
      lines = content.split("\n")
      result = {}
      comments = {}
      current_section = nil
      in_multiline_string = false
      multiline_key = nil
      multiline_content = []

      lines.each_with_index do |line, i|
        trimmed_line = line.strip

        # Handle multiline strings
        if in_multiline_string
          if trimmed_line == '"""'
            if current_section && multiline_key
              result[current_section][multiline_key] = multiline_content.join("\n")
            end
            in_multiline_string = false
            multiline_key = nil
            multiline_content = []
            next
          end
          multiline_content << line
          next
        end

        # Capture comments
        if trimmed_line.start_with?('#')
          comments[i] = trimmed_line
          next
        end

        # Skip empty lines
        next if trimmed_line.empty?

        # Section header
        section_match = trimmed_line.match(/^\[(.+)\]$/)
        if section_match
          current_section = section_match[1]
          result[current_section] = {}
          next
        end

        # Key-value pair
        if current_section && trimmed_line.include?('=')
          separator_index = trimmed_line.index('=')
          key = trimmed_line[0...separator_index].strip
          value_str = trimmed_line[(separator_index + 1)..-1].strip

          # Check for multiline string start
          if value_str == '"""'
            in_multiline_string = true
            multiline_key = key
            next
          end

          value = parse_value(value_str)
          result[current_section][key] = value
        end
      end

      [result, comments]
    end

    # Parse a TSK value string into appropriate Ruby type
    private_class_method def self.parse_value(value_str)
      # Null
      return nil if value_str == 'null'

      # Boolean
      return true if value_str == 'true'
      return false if value_str == 'false'

      # Number
      return value_str.to_i if value_str.match?(/^-?\d+$/)
      return value_str.to_f if value_str.match?(/^-?\d+\.\d+$/)

      # String
      if value_str.start_with?('"') && value_str.end_with?('"')
        return value_str[1...-1].gsub('\\"', '"').gsub('\\\\', '\\')
      end

      # Array
      if value_str.start_with?('[') && value_str.end_with?(']')
        array_content = value_str[1...-1].strip
        return [] if array_content.empty?

        items = split_array_items(array_content)
        return items.map { |item| parse_value(item.strip) }
      end

      # Object/Hash
      if value_str.start_with?('{') && value_str.end_with?('}')
        obj_content = value_str[1...-1].strip
        return {} if obj_content.empty?

        pairs = split_object_pairs(obj_content)
        obj = {}

        pairs.each do |pair|
          if pair.include?('=')
            eq_index = pair.index('=')
            key = pair[0...eq_index].strip
            value = pair[(eq_index + 1)..-1].strip
            # Remove quotes from key if present
            clean_key = key.start_with?('"') && key.end_with?('"') ? key[1...-1] : key
            obj[clean_key] = parse_value(value)
          end
        end

        return obj
      end

      # Return as string if no other type matches
      value_str
    end

    # Split array items considering nested structures
    private_class_method def self.split_array_items(content)
      items = []
      current = ''
      depth = 0
      in_string = false

      content.each_char.with_index do |ch, i|
        if ch == '"' && (i == 0 || content[i - 1] != '\\')
          in_string = !in_string
        end

        unless in_string
          depth += 1 if ch == '[' || ch == '{'
          depth -= 1 if ch == ']' || ch == '}'

          if ch == ',' && depth == 0
            items << current.strip
            current = ''
            next
          end
        end

        current += ch
      end

      items << current.strip unless current.strip.empty?
      items
    end

    # Split object pairs considering nested structures
    private_class_method def self.split_object_pairs(content)
      pairs = []
      current = ''
      depth = 0
      in_string = false

      content.each_char.with_index do |ch, i|
        if ch == '"' && (i == 0 || content[i - 1] != '\\')
          in_string = !in_string
        end

        unless in_string
          depth += 1 if ch == '[' || ch == '{'
          depth -= 1 if ch == ']' || ch == '}'

          if ch == ',' && depth == 0
            pairs << current.strip
            current = ''
            next
          end
        end

        current += ch
      end

      pairs << current.strip unless current.strip.empty?
      pairs
    end

    # Convert hash back to TSK string format
    def self.stringify(data)
      result = []

      data.each do |section, section_data|
        result << "[#{section}]"

        if section_data.is_a?(Hash)
          section_data.each do |key, value|
            result << "#{key} = #{format_value(value)}"
          end
        else
          result << "value = #{format_value(section_data)}"
        end

        result << ''
      end

      result.join("\n").strip
    end

    # Format a value for TSK string representation
    private_class_method def self.format_value(value)
      case value
      when nil then 'null'
      when true then 'true'
      when false then 'false'
      when String then "\"#{value.gsub('"', '\\"').gsub('\\', '\\\\')}\""
      when Numeric then value.to_s
      when Hash
        pairs = value.map { |k, v| "\"#{k}\" = #{format_value(v)}" }
        "{#{pairs.join(', ')}}"
      when Array
        items = value.map { |v| format_value(v) }
        "[#{items.join(', ')}]"
      else
        "\"#{value}\""
      end
    end
  end
end 